Controller Runtime + EnvTest Testing + CI Tweaks

pthomison opened the pull request:
https://pkg.go.dev/sigs.k8s.io/controller-runtime#section-readme

The controller runtime sig has a couple higher level functions/patterns that are really handy to cut out some of the more annoying parts of the go-client low level API

ejether commented:
Looks like #4 and #5 are intertwined a 'lil bit. We should straighten that out. 

* I like the encapsulation of the K8sClient and the changes there. It's similar to what I was thinking initially but I wasn't sure if it was "Go-like" so 👍 

* Love the testing. I hadn't sunk my teeth into that yet and I'm glad you did so I'd have something to stare at and understand it better

* I'm not thrilled with the way the actions are broken out at the moment. I have a weak preference for a single actions tree triggerd on a push with logic for PR / `main`. the `main.yml` is probably poorly named TBH.  I don't mind breaking out into different files and the referencing them into the single action tree.  It's a weak preference and I could be convinced.. But we should certainly clean up all the dangling comments before merging the PR

pthomison commented on pkg/kube/kube.go:
```diff
@@ -20,18 +20,32 @@ package kube
 )
 
-// Get k8s client.
+type K8sClient struct {
+	k8s client.Client
+}
+
+func NewK8sClient() *K8sClient {
+	return &K8sClient{
+		k8s: GetClient(nil),
+	}
+}
```
It does (depending on some stuff, but it enables it)! So basically I just wrapped the client.Client object in a local struct (read object), so then I can set up the core k8s functions to be methods of that struct/type. 

The functions being methods gives us 2 main things 
1) it lets our functions just assume that their client is initialized and just use the client 
2) it lets us only initialize the client once at the top level and then easily share it. 

It doesn't quite automagically once use a single client everywhere (currently I think there are a couple spots where it gets initialized), but it lets us get away from the new client per API call. The way that we will get to only using a single client will be to figure out a methodology to initialize one client at start and then share it (this is how kubebuilder takes it on), but we will get there eventually :)

pthomison replied:
Also an additional call out here, the NewK8sClient function is really just a QOL/nice to have 1-liner to get a client. An example of getting a K8SClient w/o the New function can be seen in the setupTestEnvironment function.

pthomison commented:
Def ++ to GH actions bits, was just kicking the tires a bit haha. So to summarize the CI/CD approach, we are looking for 
- Single file with all actions being derived from it (does action tree == file here?)
- Builds & Tests on PRs
- Builds, Tests, & Pushes on Merges/Pushes to Main

Any tweaks to the above?

And how do we feel about using golangci-lint in the action as well? TBH I can go any direction on linters, but just consolidated to golangci-lint since thats what we were rocking in the Makefile. It does mean we won't have "universal" linting of all of our types of resources (ie helm, scripts, other craziness that makes it way in), but I kinda think that will avoid more headaches that it would cause. But like I said, can def be convinced in any direction if there is a preference :)

pthomison commented:
Went ahead and reverted back to the single workflow strategy! Long term I think it might be handy to distinguish a couple workflows, but totally fine with this pattern!

This bad boy is starting to become a pretty hefty PR, so I'm gonna get it cleaned up && then I think it should be ready for your final 👁️ & 👍 / 👎 . 

Some additional changes just to call out:
- Added testing to the workflow
- Added workflow dependencies so the linting & testing must pass before the build is started && that the image has been pushed before the helm chart gets released (to avoid the weird gap between helm charts & their images getting released)
- Rev'd any GH actions that had complaints about Node16 (Is there any reason docker/login-action is pinned? Left this one in place, but its the last remaining Node16 action)
- Added docker image caching using the repository to push up a cache layer

LMK your thoughts @ejether ! Also more than happy to sync up via phone/meet if we wanna deep dive any of the go stuff! 🤘
